File: frontend/src/lib/game/modeRules.ts
import type { GameMode } from './types';

/**
 * 打击出现位置：
 * - direct=直接在 TargetSystem 出现并即刻判定
 * - ownerPlanet=从 owner 星球飞行到达后判定
 * - stealthOwnerPlanet=「隐逐跳」：行为同 ownerPlanet，但飞行路径仅拥有者可见；
 *   对其他玩家仅揭露 TargetSystem 与当前位置到目标的图最短跳数距离；回放可见完整路径
 */
export type StrikeOrigin = 'direct' | 'ownerPlanet' | 'stealthOwnerPlanet';

/**
 * 打击落空处理：discard=废弃到弃牌堆 / freeControl=保留可自由控制 / requireTarget=保留必须先指定新 TargetSystem
 */
export type StrikeMissBehavior = 'discard' | 'freeControl' | 'requireTarget';

/**
 * 光速飞船使用方式：oneTime=一次性(Classic)，reusable=可复用(Relics)
 */
export type LightspeedUsage = 'oneTime' | 'reusable';

/**
 * 描述特定游戏模式的规则差异。字段为编译期常量，与后端 mode_rules.go 保持一致。
 */
export interface ModeRules {
  /** 光速飞船使用方式 */
  lightspeedUsage: LightspeedUsage;
  /** Classic 合并动作成本 */
  lightspeedCombinedActionCost: number;
  /** Relics 部署成本 */
  lightspeedDeployCost: number;
  /** Relics 跃迁成本 */
  lightspeedJumpCost: number;
  /** 携带能量上限 */
  lightspeedCarryCap: number;
  /** 是否启用留言 */
  lightspeedMessageEnabled: boolean;
  /** 是否启用遗迹分布 */
  relicDistributionEnabled: boolean;
  /** 打击出现位置 */
  strikeOrigin: StrikeOrigin;
  /** 打击落空处理 */
  strikeMissBehavior: StrikeMissBehavior;
  /** 打击是否可摧毁遗迹（true=Relics 默认，false=Classic 默认） */
  strikeCanDestroyRelic: boolean;
}

const classicModeRules: ModeRules = {
  lightspeedUsage: 'oneTime',
  lightspeedCombinedActionCost: 10,
  lightspeedDeployCost: 0,
  lightspeedJumpCost: 0,
  lightspeedCarryCap: 0,
  lightspeedMessageEnabled: false,
  relicDistributionEnabled: false,
  strikeOrigin: 'direct',
  strikeMissBehavior: 'discard',
  strikeCanDestroyRelic: false,
};

const relicsModeRules: ModeRules = {
  lightspeedUsage: 'reusable',
  lightspeedCombinedActionCost: 0,
  lightspeedDeployCost: 10,
  lightspeedJumpCost: 3,
  lightspeedCarryCap: 5,
  lightspeedMessageEnabled: true,
  relicDistributionEnabled: true,
  strikeOrigin: 'ownerPlanet',
  strikeMissBehavior: 'discard',
  strikeCanDestroyRelic: true,
};

/**
 * 返回指定游戏模式的静态规则。未知模式回退到 Classic。
 *
 * 重载：接受字符串模式名（向后兼容），或包含 {gameMode, modeRules} 的状态对象。
 * 当 state.modeRules 存在时优先返回该自定义覆盖值（与后端 StateRules(state) 语义一致）。
 * 后端 GameState / ViewState 经由 CreateViewState 透传 state.ModeRules，
 * 自定义房间（房主在创建时调整规则）的对局按自定义规则渲染。
 */
export function getModeRules(input: GameMode | string | undefined): ModeRules;
export function getModeRules(input: { gameMode?: GameMode | string; modeRules?: ModeRules | null } | undefined): ModeRules;
export function getModeRules(
  input:
    | GameMode
    | string
    | { gameMode?: GameMode | string; modeRules?: ModeRules | null }
    | undefined,
): ModeRules {
  if (input && typeof input === 'object' && 'modeRules' in input) {
    // 状态对象形式：自定义规则优先于预设
    if (input.modeRules) return input.modeRules;
    return getModeRules(input.gameMode);
  }
  // 字符串形式：按 gameMode 选预设
  if (input === 'civilization_relics') return relicsModeRules;
  return classicModeRules;
}

File: mcpserver/internal/semantic/mode_rules.go
package semantic

// mode_rules.go — MCP Server 侧的模式规则类型与查询函数。
//
// 设计：
//   - ModeRules 结构体与后端 backend/internal/game/mode_rules.go 对齐。
//   - 规则常量值由 mode_rules_gen.go（代码生成产物）提供，数据源为后端 game 包，
//     确保三处文案一致。运行 "go run ../../backend/cmd/codegen" 在 mcpserver/ 下重新生成。
//   - 字符串枚举常量（StrikeOriginDirect 等）保留在此（mcpserver 独有设计）。
//   - MCP Server 是独立进程，不能依赖后端运行时；生成产物保证无后端连接时仍可工作。
//
// 与后端的差异：
//   - 后端 StrikeOrigin / StrikeMissBehavior 使用私有 iota 枚举，
//     此处改为 string 常量，便于 JSON 序列化与跨进程消费。
//   - 增加 Mode 与 Description 字段，供 Resource 消费者识别模式与理解差异。

// StrikeOriginDirect 表示打击直接在 TargetSystem 出现并即刻判定(Classic 模式)。
const StrikeOriginDirect = "direct"

// StrikeOriginOwnerPlanet 表示打击从 owner 星球出现,逐跳飞行到达 TargetSystem 后判定(Relics 模式)。
const StrikeOriginOwnerPlanet = "ownerPlanet"

// StrikeOriginStealthOwnerPlanet 表示「隐逐跳」:行为同 OwnerPlanet,
// 但飞行路径仅拥有者可见;对其他玩家仅揭露 TargetSystem 与当前位置到目标的图最短跳数距离。
// 回放(REPLAY)观察者可见完整路径。
const StrikeOriginStealthOwnerPlanet = "stealthOwnerPlanet"

// StrikeMissDiscard 表示打击落空时废弃到弃牌堆(Classic / Relics 模式)。
const StrikeMissDiscard = "discard"

// StrikeMissFreeControl 表示打击落空时保留为 Missed 飞行打击,玩家自由重定向/跳过/废弃。
const StrikeMissFreeControl = "freeControl"

// StrikeMissRequireTarget 表示打击落空时保留为 Missed 飞行打击,玩家必须指定新 TargetSystem 或废弃。
const StrikeMissRequireTarget = "requireTarget"

// ModeClassic 是经典模式标识,对齐后端 game.GameModeClassic。
const ModeClassic = "classic"

// ModeCivilizationRelics 是文明遗迹模式标识,对齐后端 game.GameModeCivilizationRelics。
const ModeCivilizationRelics = "civilization_relics"

// ModeRules 描述特定游戏模式的规则差异,镜像后端 game.ModeRules。
// 字段为编译期常量,运行时不变。
type ModeRules struct {
	// Mode 是模式标识(classic / civilization_relics)。
	Mode string `json:"mode"`

	// 光速飞船规则
	// LightspeedOneTime: true=一次性(Classic), false=可复用(Relics)
	LightspeedOneTime                     bool `json:"lightspeedOneTime"`
	LightspeedCombinedActionCost          int  `json:"lightspeedCombinedActionCost"`          // Classic 合并动作成本(random)
	LightspeedCombinedActionCostSpecified int  `json:"lightspeedCombinedActionCostSpecified"` // Classic 合并动作成本(specified)
	LightspeedDeployCost                  int  `json:"lightspeedDeployCost"`                  // Relics 部署成本
	LightspeedJumpCostRandom              int  `json:"lightspeedJumpCostRandom"`              // Relics 跃迁成本(random)
	LightspeedJumpCostSpecified           int  `json:"lightspeedJumpCostSpecified"`           // Relics 跃迁成本(specified)
	LightspeedCarryCap                    int  `json:"lightspeedCarryCap"`                    // 携带能量上限
	LightspeedMessageEnabled              bool `json:"lightspeedMessageEnabled"`              // 是否启用留言

	// 遗迹规则
	RelicDistributionEnabled bool `json:"relicDistributionEnabled"` // 是否启用遗迹分布

	// 打击规则
	StrikeOrigin          string `json:"strikeOrigin"`          // 打击出现位置: direct / ownerPlanet
	StrikeMissBehavior    string `json:"strikeMissBehavior"`    // 打击落空处理: discard / freeControl / requireTarget
	StrikeCanDestroyRelic bool   `json:"strikeCanDestroyRelic"` // 打击可否摧毁遗留物

	// Description 是模式的人类可读摘要(中文,事实陈述)。
	Description string `json:"description"`
}



// GetModeRules 按 mode 标识返回对应 ModeRules。
// 对齐前端 modeRules.ts:72-75 的回退语义：未知模式（含空串）回退到 Classic。
// mode 取值: classic / civilization_relics / 其他（回退 Classic）。
// 第二个返回值始终为 true（向后兼容签名，未来可移除）。
func GetModeRules(mode string) (ModeRules, bool) {
	switch mode {
	case ModeCivilizationRelics:
		return relicsModeRules, true
	default:
		// 包括 ModeClassic 与所有未知模式（含空串）
		return classicModeRules, true
	}
}

// ListModeRules 返回全部模式规则(2 套:Classic + Civilization Relics)。
// 用于 Resource 消费者枚举所有可用模式。
func ListModeRules() []ModeRules {
	return []ModeRules{classicModeRules, relicsModeRules}
}
